### ISSUE TITLE
Job Control triggering erroneous errors from javac

### ISSUE THREAD

**Issue by pgdouyon** [neovim/neovim#2143]
I'm toying around with a plugin to run Vim's compiler plugins asynchronously and ran into a strange bug.  I have a compiler plugin that figures out the classpath and sets makeprg to `javac -classpath <classpath> %` to run javac on the current buffer and detect syntax errors.  I ran into the following situation:

### Syntax error in the buffer ###
- `:compiler javac | silent make` correctly reports the error
- `:compiler javac | echo system(substitute(&makeprg, '\V%', expand("%"), "g"))` correctly reports the error
- A job control setup which essentially uses 
`jobstart(name, split(&makeprg)[0], split(&makeprg)[1:])` correctly reports the error

### No syntax error in the buffer ###
- `:compiler javac | silent make` correctly reports no error
- `:compiler javac | echo system(substitute(&makeprg, '\V%', expand("%"), "g"))` correctly reports no error
- A job control setup which essentially uses 
`jobstart(name, split(&makeprg)[0], split(&makeprg)[1:])` incorrectly reports hundreds of missing symbols/classes which seems to imply that the classpath is wrong/corrupted.

I can't really explain this behavior and don't know how to test it anymore so I was hoping someone could help me figure out what's going on.  The job control setup wasn't doing anything more complicated than calling jobstart and then `echom join(v:job_data[2], "\n")`.  

My initial guess would have been that the length of the classpath is causing problems (length is 15,623 - java classpaths get ridiculous), but that doesn't explain why the job control setup runs as expected when a syntax error is present in the buffer.

I can provide a minimum example to reproduce, but that would still need to be a fairly large example so I figured I would just start by proposing the problem and go from there.

OS: Mac OS X 10.10.2
Terminal: iTerm2

**Comment by justinmk** [comment]
Note that `jobstart()` does *not* do shell expansion, so make sure you don't have any wildcards in the arguments or escaped characters.

**Comment by pgdouyon** [comment]
I don't have any wildcards, the directories are all full paths to jar files.  I also tried `shellescape()` on the arguments but it makes no difference.  I don't believe `system()` expands wildcards either if I'm not mistaken, that's the reason I tested it with that function.

On the other hand, I just realized like an idiot that the reason job control doesn't have problems with a syntax error present in the buffer is most likely because javac scans for syntax errors before attempting to resolve symbols.  Which then makes me wonder if job-control is indeed cropping the extremely long classpath.  Since it's an isolated entry in the arg list, it could be cropped without corrupting the other arguments.

Is there some kind of limit on the length of arguments we can pass to a job control process?

**Comment by pgdouyon** [comment]
Sorry about that, just confirmed that if I move the classpath argument to the $CLASSPATH environment variable (removing the 15,000 character argument) the problems go away.  So I would assume that Neovim does crop the arguments.  If that's the case, can I propose that the limits Neovim places on arguments to job control match the limits on `:make` and `system()` and that it be mentioned somewhere in the help doc?

**Comment by justinmk** [comment]
> if job-control is indeed cropping the extremely long classpath.

Oops, I just noticed your 15,623 number. Yes, this is a known issue with the current RStream logic: https://github.com/neovim/neovim/issues/1813


> I don't believe system() expands wildcard

`system()` *does* expand wildcards, it is executed in a shell context (jobstart() is not). `system()` is affected by the Vim `&shell`, `&shellcmdflag`, etc., functions (whereas `jobstart()`) is not. 

P.S. Just for reference, `shellescape()` is exactly the opposite of what `jobstart()` wants. `jobstart()` does its own "escaping": each argument is passed as a literal. So if you call `jobstart("foo", "bar", "*")`, the literal `*` character will be sent to `bar`, it will not be expanded. If you call `jobstart("foo", "bar", "\*")`, then the literal `\*` string will be sent to `bar`.

**Comment by justinmk** [comment]
> mentioned somewhere in the help doc?

The current limit is a bug. But I do have an unfinished update to the documentation for job control.

**Comment by pgdouyon** [comment]
Ahhh okay I understand now, I hadn't seen that issue before.  Since the problem seems to be solely caused by a known bug which already has an issue assigned, I'm fine with closing this one then if that's okay with you.  I don't really want to pollute the issue tracker.

**Comment by justinmk** [comment]
Let's leave it open for now.

**Comment by tarruda** [comment]
what happens if you use:

```vim
call jobstart(name, &p_sh, ['-c', substitute(&makeprg, '\V%', expand("%"), "g")])
```

**Comment by pgdouyon** [comment]
Sorry for the delay, but the above workaround also solves the problem, even with the 15,000 char classpath.  Although, I don't know what the &p_sh variable is, neovim printed an error when I tried to reference it.  I just replaced it with "bash" and everything worked as expected, not sure if that's what you meant.

**Comment by tarruda** [comment]
@pgdouyon its actually `&sh`, I confused with the internal variable name :smile: 

When you run a program with `system` or `!`, Neovim is simply calling your shell and passing the command to the `-c` argument.

If that works, then it means that javac relies on your shell to set the classpath, which is why you can't spawn it directly with `jobstart`.

**Comment by pgdouyon** [comment]
Sorry I'm still a bit confused, why would the classpath be set any different in

`call jobstart(name, 'javac', ['-classpath', 'really-long-classpath', 'current-file'])`
versus
`call jobstart(name, &sh, ['-c', 'javac -classpath really-long-classpath current-file'])`

In both cases the `$CLASSPATH` variable is unset and so I would think javac should be using the classpath passed to it as argument.  (although setting `$CLASSPATH` does allow me to run `call jobstart(name, 'javac', ['current-file'])` without any problems which is the current solution I'm using).

Are you suggesting that the shell somehow parses out the classpath in the second example, or does this have to do more with how job-control passes arguments to job processes?


** I'm still curious about the above question, but I think I've actually accidentally found the source of the problem **
In trying to reproduce the above example with the two different calls to jobstart() I found that I couldn't reproduce the error.  After some poking around, I realized this was because I had made some changes to my compiler script where I was no longer quoting the classpath value.  Whenever I do quote the classpath value (whether single or double quotes) the errors occur.  So in summary:

#### Works as expected ####
`call jobstart(name, 'javac', ['-classpath', 'really-long-classpath', 'current-file'])`
`call jobstart(name, &sh, ['-c', 'javac -classpath really-long-classpath current-file'])`

#### Throws erroneous errors ####
`call jobstart(name, &sh, ['-c', 'javac -classpath 'really-long-classpath' current-file'])`
`call jobstart(name, &sh, ['-c', 'javac -classpath "really-long-classpath" current-file'])`
`call jobstart(name, 'javac', split('-classpath "really-long-classpath" current-file'))`
`call jobstart(name, 'javac', split("-classpath 'really-long-classpath' current-file"))`

which should be equivalent to:
`call jobstart(name, 'javac', ['-classpath', '"really-long-classpath"', 'current-file'])`
`call jobstart(name, 'javac', ['-classpath', '''really-long-classpath''', 'current-file'])`

When running javac with a quoted classpath on the command line it works as expected and doesn't throw the errors.

**Comment by pgdouyon** [comment]
I would assume that this is because job-control does it's own escaping of arguments, arguments shouldn't be quoted when passed in.  Otherwise I'd assume they get quoted twice and are parsed incorrectly.  I'm not sure what the reasoning is for job-control only passing in literal arguments, but maybe it could check if arguments are already quoted before escaping them?

**Comment by tarruda** [comment]
@pgdouyon I know very little about `javac` so I can't answer your question, what I can tell you is how job control and other process-spawning facilities work in Neovim:

- Bang(`!`), `system`, `make` and every other command that spawns processes(except `jobstart`/`rpcstart`) passes the command string to the shell. So all shell quoting/escape rules apply
- `jobstart` is a lower level function that uses operating system calls to spawn the processes. Thats why you pass a list as arguments(the prog and arguments are directly translated to the `char **argv` passed to the spawned program

In other words, shell initialization files(that set `$CLASSPATH` for example) wont have any effect when you use `jobstart`. Quotes and spaces are also shell specific concerns, and consequently have no effect with `jobstart`.

For example, if you do:

```vim
call jobstart('str', 'prog', ['-c "abc"', '-d'])
```

the `char **argv` parameter passed to main will contain `prog`, `-c "abc"` and `-d`, where for

```vim
call system('prog -c "abc" -d')
```

it will be `prog`, `-c`, `abc` and `-d`(because the whole string is passed to the shell, which parses/splits the command string)

**Comment by justinmk** [comment]
> I'm not sure what the reasoning is for job-control only passing in literal arguments, 

For the escaping circus (which differs across shells and OSes) there is `system()` and `:!`. It's really quite nice to have a mechanism that is guarnteed to not try any of that "magic" (which is the source of numerous bugs and painful discussions on vim_dev).

> but maybe it could check if arguments are already quoted before escaping them?

Keeping it simple (i.e., literal, no magic) has advantages. For wildcard expansion there is `glob()` and `expand()`.

**Comment by pgdouyon** [comment]
Okay I think I understand now.  If job-control is intended to be a lower-level, shell-agnostic protocol I agree it makes sense that it shouldn't be handling quoted arguments, and they shouldn't even be passed to it since argument quoting is handled by shell programs.  

I can definitely see the advantages of it now, and I think it was confusing for me before because I was envisioning jobstart() as an asynchronous system() which is still executed in some kind of shell process where you could pass a command the way you'd type it on the command line.

Unless job-control should be handling quoted strings, then I don't really think there's a problem/bug here, just a misunderstanding of how job-control should be used.  Passing the classpath unquoted causes no errors, even though it's extremely long.  This explanation was really helpful though and I definitely think it'd be nice if we put this in the help docs in some form (I know the docs are far from finished).  Just something mentioning that job-control is a low level, shell-agnostic process and it's arguments should be unquoted and manually expanded so that people don't confuse it with how system() behaves in Vim.

Also, regarding javac, the classpath just indicates to javac which directories/jar files it can find imported classes in so that it's able to properly resolve them.  If the classpath isn't passed on the command line via the '-classpath' option then javac expects to find it in the $CLASSPATH environment variable.  I was setting the environment variable in the compiler plugin in vim (not in a shell initialization file), i.e. `let $CLASSPATH = s:parse_classpath(current-file)`.  I don't know a lot about how job-control handles environment variables, but since setting the environment variable in Neovim still allowed javac to find it I assume that the job-control process somehow inherits all the environment variables within the scope of the Neovim process it was executed from.

Thanks for all your help though, it's been really fun working with job-control and getting async syntax/style error updates working in Neovim (even though all my coworkers still judge me for not using Eclipse/IntelliJ).

**Comment by tarruda** [comment]
>  I was envisioning jobstart() as an asynchronous system() which is still executed in some kind of shell process where you could pass a command the way you'd type it on the command line.

As I said before, job control can be used to emulate an asynchronous `system`, simply pass `&sh` as the program and `['-c', 'CMD']` as the arguments. This is essentially what `system` does under the hoods(everything is implemented on top of the job control module)

> Thanks for all your help though, it's been really fun working with job-control and getting async syntax/style error updates working in Neovim (even though all my coworkers still judge me for not using Eclipse/IntelliJ).

Thanks for working to make Neovim better for java development. Soon it will be possible to integrate command-line debuggers such as `jdb` which will certainly improve the IDE experience.
